### DISCUSSION TITLE
`onBrokenMarkdownLinks` vs `onBrokenLinks`

### DISCUSSION THREAD

**Discussion by pokey** [facebook/docusaurus#8613] [Q&A]
How come there are both `onBrokenMarkdownLinks` and `onBrokenLinks`?  When should one be using the former vs the latter?

I had `onBrokenMarkdownLinks` set to `throw`, but had to drop it to `warn` because I had a custom transform to patch up a relative link that was [output incorrectly](https://github.com/tgreyuk/typedoc-plugin-markdown/issues/394) by `docusaurus-plugin-typedoc`.  The `onBrokenMarkdownLinks` checker seems to run before my custom transform, so it flags the link, even though the custom transform is registered in `beforeDefaultRemarkPlugins`

I was surprised to discover that `onBrokenLinks` seems to catch the same problem, so it seems to offer just as much protection against bad links, while allowing for rewrites.  So I guess my questions are:

1. Why would I use `onBrokenMarkdownLinks` at all?  Are there things it catches that `onBrokenLinks` doesn't?
2. Why does `onBrokenMarkdownLinks` run before my rewrites, even though they're defined in `beforeDefaultRemarkPlugins`?

Happy to file a PR to update the docs once I understand this stuff; the docs are pretty light on these config options

**Comment by slorber** [comment]
`onBrokenMarkdownLinks` runs earlier and will tell you if you reference markdown files that do not exist in links such as `[link](./doesNotExist.md)`. It only reports linking errors found in md docs, not any other React code.

`onBrokenLinks` is more generic, runs at the very end once we have rendered your whole site, and will tell you if there's any invalid internal link in your site (for now it does not check anchor links), including bad links from React pages, third-party themes, bad themeConfig...

> The onBrokenMarkdownLinks checker seems to run before my custom transform, so it flags the link, even though the custom transform is registered in beforeDefaultRemarkPlugins

That's not supposed to be the case, please double check and provide a repro but I'm pretty sure you are wrong here. Note that files are processed in parallel so remark1 of file1 could run after remarkplugin2 of file2.

> Why would I use onBrokenMarkdownLinks at all? Are there things it catches that onBrokenLinks doesn't?

It catches broken markdown links earlier in the process and could give you a taster feedback. Our most reliable broken link checker runs at the very end and it's a quote slow feedback, that's why there's another one running before. I can understand it's confusing at the same time we had to expose this option 😅 

---

> Happy to file a PR to update the docs once I understand this stuff; the docs are pretty light on these config options

Would be happy to improve our docs. Please note that I'd rather keep the doc simple and not explain too many implementation details. It's possible that we'll improve our broken link checker in the future to provide a quicker feedback loop. The most important thing users should understand is that one `onBrokenLinks` is the most reliable security as it actually "runs" the code and inspects the output. The other is limited to what we can be checked at markdown processing time, and will not report invalid React page links.

**Reply by pokey** [reply]
Got it thanks for the helpful response!  So it sounds like it's safe for me to just drop the markdown link checker for now

> please double check and provide a repro

Ok I'll see if I can get a minimal repro for you here.  The reason I believe this to be the case is that when I have `onBrokenMarkdownLinks` set to `throw`, I get the broken link error and my transform never runs, but if it's set to `warn`, the transform runs and the `onBrokenLinks` check is happy with the transformed output.  But there's enough moving pieces here that I'll acknowledge that there maybe something I'm missing here 😅.  Could be related to the fact that the markdown output is generated by `docusaurus-plugin-typedoc`?

**Reply by slorber** [reply]
👍 

> The reason I believe this to be the case is that when I have onBrokenMarkdownLinks set to throw, I get the broken link error and my transform never runs

If the site fails on the very first markdown file, then it may stop processing files even before it process file2 which your plugin would handle. Maybe just log that your plugin runs first before just logging the successful processings?

Double checked and I don't see anything wrong with the order of our code:

![CleanShot 2023-02-01 at 17 49 58@2x](https://user-images.githubusercontent.com/749374/216108360-99038308-e886-44b2-837b-8fb112925f7c.png)
